src/commands/signzone.rs

```diff
@@ -1276,13 +1288,59 @@ impl SignZone {
                     for covering_rrsigs in owner_rrs
                         .rrsets()
                         .filter(|this_rrset| this_rrset.rtype() == Rtype::RRSIG)
-                        .map(|this_rrset| this_rrset.iter().filter(|rr| matches!(rr.data(), ZoneRecordData::Rrsig(rrsig) if rrsig.type_covered() == rrset.rtype())))
+                        .map(|this_rrset| {
+                            this_rrset.iter().filter(|rr| {
+                                matches!(rr.data(), ZoneRecordData::Rrsig(rrsig)
+                                    if rrsig.type_covered() == rrset.rtype()
+                                        && if self.invoked_as_ldns && rr.owner() == apex {
+                                            // Withhold an RRSIG that covers the NSEC of the apex
+                                            // as we' reinserting them at the end of the apex' RRsets
+                                            rrsig.type_covered() != Rtype::NSEC
+                                        } else { true }
+                                )
+                            })
+                        })
+                    {
+                        for covering_rrsig_rr in covering_rrsigs {
+                            self.writeln_rr(&mut writer, covering_rrsig_rr)?;
+                        }
+                    }
+                }
+
+                // If running as ldns-signzone, we've been withholding the NSEC and NSEC's RRSIG at
+                // the apex above to reinsert them after all other RRsets at the apex. By default,
+                // the DNSKEY RRset and it's RRSIG would take the rear of the RRsets at the apex.
+                // This doesn't apply, if we're using NSEC3. Additionally, the NSEC RRs at other
+                // places than the apex do not have the ordering issue.
+                if self.invoked_as_ldns && !self.use_nsec3 && owner_rrs.owner() == apex {
+                    for covering_nsec in owner_rrs
+                        .rrsets()
+                        .filter(|this_rrset| this_rrset.rtype() == Rtype::NSEC)
+                        .map(|this_rrset| this_rrset.iter())
+                    {
+                        for covering_nsec_rr in covering_nsec {
+                            self.writeln_rr(&mut writer, covering_nsec_rr)?;
+                        }
+                    }
```

Comment:
This could be simplfied by using `find()` and `first()` instead of `filter()` as there is only ever one NSEC RR in an RRset, e.g.:

```suggestion
                    if let Some(nsec_rrset) = owner_rrs
                        .rrsets()
                        .find(|this_rrset| this_rrset.rtype() == Rtype::NSEC)
                    {
                        self.writeln_rr(&mut writer, nsec_rrset.first())?;
                    }
```